esp-hal-common/src/lib.rs

```diff
@@ -202,25 +202,38 @@ mod critical_section_impl {
 
     #[cfg(xtensa)]
     mod xtensa {
+
+        #[cfg(multi_core)]
+        use super::multicore::{LockKind, MULTICORE_LOCK};
+
         unsafe impl critical_section::Impl for super::CriticalSection {
             unsafe fn acquire() -> critical_section::RawRestoreState {
-                let tkn: critical_section::RawRestoreState;
+                let mut tkn: critical_section::RawRestoreState;
                 core::arch::asm!("rsil {0}, 5", out(reg) tkn);
                 #[cfg(multi_core)]
                 {
-                    let guard = super::multicore::MULTICORE_LOCK.lock();
-                    core::mem::forget(guard); // forget it so drop doesn't run
+                    match super::multicore::MULTICORE_LOCK.lock() {
+                        LockKind::Lock => {}
+                        LockKind::Reentry => tkn |= 1 << 31,
+                    }
                 }
                 tkn
             }
 
-            unsafe fn release(token: critical_section::RawRestoreState) {
+            unsafe fn release(mut token: critical_section::RawRestoreState) {
                 if token != 0 {
                     #[cfg(multi_core)]
                     {
                         debug_assert!(super::multicore::MULTICORE_LOCK.is_owned_by_current_thread());
-                        // safety: we logically own the mutex from acquire()
-                        super::multicore::MULTICORE_LOCK.force_unlock();
+
+                        let lock_kind = if token & (1 << 31) != 0 {
+                            token = token & !(1 << 31);
+                            LockKind::Reentry
+                        } else {
+                            LockKind::Lock
+                        };
+
+                        super::multicore::MULTICORE_LOCK.unlock(lock_kind);
                     }
                     core::arch::asm!(
```

Comment:
Maybe we could add a `debug_assert` here to ensure that bits 19-31 are zeroed before we restore? Probably do the same for RISCV too.